File: src/main/java/com/refinedmods/refinedstorage/blockentity/craftingmonitor/ICraftingMonitor.java
package com.refinedmods.refinedstorage.blockentity.craftingmonitor;

import com.refinedmods.refinedstorage.api.autocrafting.ICraftingManager;
import com.refinedmods.refinedstorage.api.autocrafting.task.ICraftingTask;
import com.refinedmods.refinedstorage.blockentity.data.BlockEntitySynchronizationParameter;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;

import javax.annotation.Nullable;
import java.util.Collection;
import java.util.Optional;
import java.util.UUID;

public interface ICraftingMonitor {
    int TABS_PER_PAGE = 7;

    Component getTitle();

    void onCancelled(ServerPlayer player, @Nullable UUID id);

    BlockEntitySynchronizationParameter<Integer, ?> getRedstoneModeParameter();

    Collection<ICraftingTask> getTasks();

    @Nullable
    ICraftingManager getCraftingManager();

    boolean isActiveOnClient();

    void onClosed(Player player);

    Optional<UUID> getTabSelected();

    int getTabPage();

    void onTabSelectionChanged(Optional<UUID> taskId);

    void onTabPageChanged(int page);

    int getSlotId();
}

File: src/main/kotlin/com/refinedmods/refinedstorage/tile/craftingmonitor/ICraftingMonitor.kt
package com.refinedmods.refinedstorage.tile.craftingmonitor

import com.refinedmods.refinedstorage.api.autocrafting.ICraftingManager
import com.refinedmods.refinedstorage.api.autocrafting.task.ICraftingTask
import com.refinedmods.refinedstorage.tile.data.TileDataParameter
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.util.text.Text
import java.util.*

interface ICraftingMonitor {
    val title: Text
    fun onCancelled(player: ServerPlayerEntity?, @Nullable id: UUID?)
    val redstoneModeParameter: TileDataParameter<Int?, *>?
    val tasks: Collection<ICraftingTask?>?

    @get:Nullable
    val craftingManager: ICraftingManager?
    val isActiveOnClient: Boolean
    fun onClosed(player: PlayerEntity?)
    val tabSelected: Optional<UUID?>
    val tabPage: Int
    fun onTabSelectionChanged(taskId: Optional<UUID?>)
    fun onTabPageChanged(page: Int)
    val slotId: Int

    companion object {
        const val TABS_PER_PAGE = 7
    }
}